src/radical/pilot/utils/serializer.py

```diff
@@ -0,0 +1,154 @@
+
+"""
+The serializer should be able to (de)serialize information that we want
+to send over the wire from the client side to the agent side via
+1- ZMQ
+2- MongoDB
+
+we except:
+    1- Callables with and without dependecies.
+    2- Non-callables like classes and other python objects
+"""
+__copyright__ = "Copyright 2013-2016, http://radical.rutgers.edu"
+__license__   = "MIT"
+
+import os
+import dill
+import pickle
+import codecs
+import tempfile
+
+
+_obj_dir       = tempfile.gettempdir()
+_obj_file_name = 'rp_obj.pkl'
+_obj_file_path = os.path.join(_obj_dir, _obj_file_name)
+
+
+def serialize_obj(obj):
+    """
+    serialize object
+    """
+    result    = None
+    exception = None
+
+    if callable(obj):
+        try:
+            result = dill.dumps(obj)
+        except Exception as e:
+            exception = e
+            pass
+
+        # if we fail, then pikle it by reference
+        if result is None:
+            # see issue: https://github.com/uqfoundation/dill/issues/128
+            try:
+                result = dill.dumps(obj, byref = True)
+            except Exception as e:
+                exception = e
+                pass
+
+    else:
+        try:
+            result = dill.dumps(obj, recurse = True)
+        except Exception as e:
+            exception = e
+            pass
+
+        if not result:
+            # see issue: https://github.com/uqfoundation/dill/issues/128
+            try:
+                result = dill.dumps(obj, byref = True)
+            except Exception as e:
+                exception = e
+                pass
+
+    if result is None:
+        raise Exception("object %s is not serializable") from exception
+
+    return  result
+
+
+def serialize_file(obj):
+    """
+    serialize object to file
+    # FIXME: assign unique path and id for the pickled file
+    #        to avoid overwriting to the same file
+    """
+    result    = None
+    exception = None
+
+    if callable(obj):
+        try:
+            with open(_obj_file_path, 'wb') as f:
+                dill.dump(obj, f)
+                result = _obj_file_path
+
+        except Exception as e:
+            exception = e
+            pass
```

Comment:
@mtitov  You are correct about the `pass`, but for the exception, you must keep it somewhere to raise from it in the end.
Also, we do not want the serializer to stop on one way of serialization (in case it fails). We continue to try to serialize with different approaches until all of them fail then we raise from the last try. Hope this was clear.

Reply:
> for the exception, you must keep it somewhere to raise from it in the end

yeah, it will stay in `e`

> we do not want the serializer to stop on one way of serialization (in case it fails)

right, I didn't mean that it will be called before other checks
```python
...
        try:
            with open(_obj_file_path, 'wb') as f:
                dill.dump(obj, f)
        except Exception as e:
            pass
        else:
            result = _obj_file_path
...
    if result is None:
        raise Exception('object is not serializable') from e
```